llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8391,135 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+// a unknown (run-time) index into an array, we parse the operands of the corresponding access chain, add a chain of
+// ANDed compares to check if the bounds requirements are met and then do a conditional jump to either the intermediate
+// block which handles the memory operation or the final block (which will skip the memory op, if bounds checks failed).
+// As mentioned, this does preparation work like inserting the bounds checks, its up to the caller caring about
+// inserting the OpLoad or OpStore. This also sets a flag that signals the main translation loop that we want to
+// continue inserting subsequent instructions at the final block instead of just going back to the parent block.
+// Currently, OpLoads will yield zero and OpStores will just omit the store, if bounds checks failed. This is up to the
+// caller. This implementation handles nested structs using arrays and 1...n-dimensional arrays. This returns if we need
+// to insert a bounds checks and pointers to the newly created basic blocks (which can be optimized away by subsequent
+// passes, if not needed).
+//
+// An example: We have the following sequence in LLVM IR:
+//
+// entry:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+//
+// which can cause a out of bounds access.
+//
+// However, we are converting it to a sequence like this:
+//
+// entry:
+//
+// %shouldLoad1 = icmp ult i32 runtimeIndex, localArray.dim1.upperBound
+// %shouldLoad2 = icmp ult i32 runtimeIndex * 2, localArray.dim2.upperBound
+// %shouldLoad = and i1 %shouldLoad1, %shouldLoad2
+// br i1 %shouldLoad, label %loadBlock, label %finalBlock
+//
+// loadBlock:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+// br label %finalBlock
+//
+// finalBlock:
+// %finalResult = phi ... <4 x float> [ zeroinitializer, %entry], [ %load, loadBlock ]
+// ... use %finalResult subsequently
+//
+SPIRVToLLVM::ScratchBoundsCheckInfo SPIRVToLLVM::prepareScratchBoundsCheck(SPIRVValue *toBeWrapped) {
+  constexpr ScratchBoundsCheckInfo nullValue{};
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return nullValue;
+  }
+
+  if (!toBeWrapped || toBeWrapped->getOpCode() != OpAccessChain) {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  // Fetch the access chain behind the pointer the memory operation uses.
+  const auto accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+  const auto accessChainIndices = accessChainBase->getIndices();
+  if (accessChainIndices.empty()) {
+    return nullValue;
+  }
+
+  SPIRVType *baseType = accessChainBase->getBase()->getType()->getPointerElementType();
+
```

Comment:
```suggestion
if (!baseType) {
  return nullValue;
{
```
Hi, just a small suggestion here, please add NULL pointer check to avoid static code analysis warning